File: hello-fory-go/hello_fory.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	fory "github.com/apache/fory/go/fory"
)

type UserInfo struct {
	UserId     int64
	Name       string
	Age        int32
	Emails     []string
	Properties map[string]string
	Scores     []int32
	Active     bool
	CreatedAt  int64
	Avatar     []byte
}

func createSampleUserInfo() *UserInfo {
	return &UserInfo{
		UserId: 12345,
		Name:   "John Doe",
		Age:    30,
		Emails: []string{"john.doe@example.com", "johndoe@gmail.com"},
		Properties: map[string]string{
			"city":    "San Francisco",
			"country": "USA",
			"role":    "Developer",
		},
		Scores:    []int32{85, 90, 78, 92},
		Active:    true,
		CreatedAt: time.Now().Unix(),
		Avatar:    []byte{0, 1, 2, 3, 4, 5},
	}
}

func deserializeUser(f *fory.Fory, filePath string) (*UserInfo, error) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Recovered from panic: %v\n", r)
		}
	}()

	serialized, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	var userInfo UserInfo
	if err := f.Unmarshal(serialized, &userInfo); err != nil {
		return nil, err
	}

	return &userInfo, nil
}

func main() {
	// Create output directory if it doesn't exist
	outputDir := "../workspace"
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		fmt.Printf("Failed to create output directory: %v\n", err)
		os.Exit(1)
	}

	// Create a sample UserInfo object
	userInfo := createSampleUserInfo()
	fmt.Printf("Created user: %+v\n", userInfo)

	// Create fory instance configured for cross-language compatibility
	f := fory.NewFory(true)
	f.SetLanguage(fory.XLANG)
	err := f.RegisterNamedType(UserInfo{}, "org.feuyeux.fory.UserInfo")
	if err != nil {
		return
	}

	// Serialize the user to bytes
	serialized, err := f.Marshal(userInfo)
	if err != nil {
		fmt.Printf("Failed to serialize: %v\n", err)
		os.Exit(1)
	}

	// Write the serialized data to a file
	outputFile := filepath.Join(outputDir, "userinfo_go.fory")
	if err := os.WriteFile(outputFile, serialized, 0644); err != nil {
		fmt.Printf("Failed to write file: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Serialized to %s\n", outputFile)

	// Try to read serialized data from other languages
	langs := []string{"java", "go", "js", "python", "rust"}
	for _, lang := range langs {
		langFile := filepath.Join(outputDir, fmt.Sprintf("userinfo_%s.fory", lang))
		if _, err := os.Stat(langFile); err == nil {
			if userInfoFromFile, err := deserializeUser(f, langFile); err != nil {
				fmt.Printf("Failed to deserialize [%s]: %v\n", lang, err)
			} else {
				fmt.Printf("Deserialized [%s] user info from %s:%+v\n", lang, langFile, userInfoFromFile)
			}
		}
	}
}

File: hello-fory-rust/hello_fory.rs
use fory::{Error, Fory};
use fory::ForyObject;
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(ForyObject, Debug, Clone)]
struct UserInfo {
    user_id: Option<i64>,
    name: Option<String>,
    age: Option<i32>,
    emails: Option<Vec<String>>,
    properties: Option<HashMap<String, String>>,
    scores: Option<Vec<i32>>,
    active: Option<bool>,
    created_at: Option<i64>,
    avatar: Option<Vec<i8>>,
}

fn create_sample_user_info() -> UserInfo {
    let mut properties = HashMap::new();
    properties.insert("city".to_string(), "San Francisco".to_string());
    properties.insert("country".to_string(), "USA".to_string());
    properties.insert("role".to_string(), "Developer".to_string());

    let created_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;

    UserInfo {
        user_id: Some(12345),
        name: Some("John Doe".to_string()),
        age: Some(30),
        emails: Some(vec![
            "john.doe@example.com".to_string(),
            "johndoe@gmail.com".to_string(),
        ]),
        properties: Some(properties),
        scores: Some(vec![85, 90, 78, 92]),
        active: Some(true),
        created_at: Some(created_at),
        avatar: Some(vec![0i8, 1, 2, 3, 4, 5]),
    }
}

fn deserialize_user(fory: &Fory, file_path: &Path) -> Result<UserInfo, Error> {
    let serialized = fs::read(file_path).expect("Failed to read file");
    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
        fory.deserialize::<UserInfo>(&serialized)
    }));
    
    match result {
        Ok(Ok(user_info)) => Ok(user_info),
        Ok(Err(e)) => Err(e),
        Err(_) => Err(Error::InvalidData("Panic during deserialization".into())),
    }
}

fn main() -> Result<(), Error> {
    // Create output directory if it doesn't exist
    let output_dir = Path::new("../workspace");
    fs::create_dir_all(output_dir).expect("Failed to create output directory");

    // Create a sample UserInfo object
    let user_info = create_sample_user_info();
    println!("Created user: {:?}", user_info);

    // Create fory instance configured for cross-language compatibility
    let mut fory = Fory::default()
        .xlang(true)
        .compatible(true);
    fory.register_by_namespace::<UserInfo>("org.feuyeux.fory", "UserInfo")?;

    // Serialize the user to bytes
    let serialized = fory.serialize(&user_info)?;

    // Write the serialized data to a file
    let output_file = output_dir.join("userinfo_rust.fory");
    fs::write(&output_file, &serialized).expect("Failed to write file");
    println!("Serialized to {:?}", output_file);

    // Try to read serialized data from other languages
    let langs = ["java", "go", "js", "python", "rust"];
    for lang in &langs {
        let lang_file = output_dir.join(format!("userinfo_{}.fory", lang));
        if lang_file.exists() {
            match deserialize_user(&fory, &lang_file) {
                Ok(user_info_from_file) => {
                    println!("Deserialized [{}] user info from {:?}:{:?}", lang, lang_file, user_info_from_file);
                },
                Err(e) => println!("Failed to deserialize [{}]: {:?}", lang, e),
            }
        }
    }

    Ok(())
}